### PR TITLE
feat(prefilter): add Tool_prefilter module — TF-IDF cosine similarity (#4574)

### PR DESCRIPTION
**Pull request by jeong-sik**
## Summary

TF-IDF cosine similarity 기반 tool relevance scoring 모듈.
Python 벤치마크에서 42.9% → 91.7% tool selection accuracy 증명 후 OCaml 구현.

- `lib/tool_prefilter.ml` — stateless TF-IDF + synonym dictionary (130줄)
- `test/test_tool_prefilter.ml` — 14 unit tests (recall, synonym, zero-result, edge cases)
- 프로덕션 코드 변경 없음. 통합은 후속 PR.

## Product impact

MASC keeper/agent의 tool selection accuracy를 개선하기 위한 기반 모듈.
이 PR 자체는 프로덕션 경로에 연결되지 않으므로 사용자 영향 없음.
후속 PR에서 OAS guardrails 경유로 통합 시 tool calling 신뢰도 향상.

## Evidence

| Model | Standalone | TF-IDF top-3 + LLM | Improvement |
|-------|-----------|-------------------|-------------|
| Qwen3.5-9B | 42.9% | 66.7% | +24pp |
| Qwen3.5-27B | 60.7% | 79.2% | +19pp |
| Qwen3.5-35B-A3B | 42.9% | 91.7% | +49pp |

Python 벤치마크 코드: `me` repo `feat/tool-calling-benchmark` branch.
28 test cases, essential tier 21 tools, tool_choice=required, temperature=0.0.

## Review evidence

- `dune build --root .` 성공
- 14/14 unit tests 통과 (0.007s)
- Full test suite regression 없음 (307s, tool matrix 포함)
- 프로덕션 코드 변경 0 — lib/dune, test/dune에 모듈 등록만

## Linked issue

Closes #4574

## API

```ocaml
val filter : tools:tool_schema list -> query:string -> k:int -> tool_schema list
val filter_with_scores : ... -> (tool_schema * float) list
```

Zero-result contract: `[]` on empty/disjoint query. Caller fallback 책임.

## Test plan

- [x] `dune build` 성공
- [x] 14/14 unit tests 통과
- [x] Full test suite regression 없음

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/tool_prefilter.ml [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+(* Public API                                                       *)
+(* ================================================================ *)
+
+let filter_with_scores ~(tools : Types.tool_schema list) ~(query : string)
+    ~(k : int) : (Types.tool_schema * float) list =
+  let query_tokens = tokenize query in
+  if query_tokens = [] then []
+  else
+    let docs = List.map build_document tools in
+    let idf = compute_idf (query_tokens :: docs) in
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply lib/tool_prefilter.ml]: Valid point. The IDF includes query tokens in the corpus, which slightly inflates query term IDF. In practice the corpus is 300+ tool descriptions so the effect is negligible. Standard TF-IDF uses a fixed corpus — this is a single-query runtime scorer, not a search index.
